decentraland/redux/middlewares.js

```diff
@@ -0,0 +1,60 @@
+import { applyMiddleware } from "redux";
+import {
+  SET_ACCOUNT,
+  SET_ACCOUNT_CREATION_STATUS,
+  UPDATE_ACTION_FOR_ACCOUNT_CREATION_STATUS,
+  SET_ACCOUNT_CREATION_ACTIONS,
+  ADD_TO_MY_ASSETS_LIST,
+  REMOVE_MY_ASSET_FROM_LIST,
+  SET_MY_ASSETS_LIST,
+} from "./actions";
+import {
+  storeAccount,
+  storeAccountCreationStatus,
+  storeAccountCreationActions,
+  storeMyAssets,
+} from "@helpers/storage";
+
+const storer = store => next => async action => {
```

Comment:
async middleware for a sync reducer is fine then, right?

Reply:
Yes, a middleware can return a `Promise` (Ref: https://redux.js.org/advanced/middleware#seven-examples > vanillaPromise)
About promises error handler: 

```javascript
**
 * Lets you dispatch promises in addition to actions.
 * If the promise is resolved, its result will be dispatched as an action.
 * The promise is returned from `dispatch` so the caller may handle rejection.
 */
```

Reply:
Okay, cool.

So will we await the dispatch of these actions from now on?

On the error handling topic, I'll note that in their example they use the promise before the action itself, and here we're doing the opposite. Which I think is okay - I'm pretty sure our returning the value from the async function after dispatching the sync action will work too.

Reply:
Hmmm
On the example, I've understood that since the middleware is returning a `Promise` we can write our middleware function as `async` too. And the example's action is a promise, we can do that from now but none of our actions are async. Make sense?

Reply:
> On the example, I've understood that since the middleware is returning a `Promise` we can write our middleware function as `async` too.

> And the example's action is a promise, we can do that from now but none of our actions are async. Make sense?

I would say that the example uses a promise in the middleware, but the action itself is not a promise. Note that the comment above the vanilla example says, "If the promise is resolved, its result will be dispatched as an action."

Reply:
Here is the full example:

```javascript
/**
 * Lets you dispatch promises in addition to actions.
 * If the promise is resolved, its result will be dispatched as an action.
 * The promise is returned from `dispatch` so the caller may handle rejection.
 */
const vanillaPromise = store => next => action => {
  if (typeof action.then !== 'function') {
    return next(action)
  }

  return Promise.resolve(action).then(store.dispatch)
}
```

Seems to me that the action is async since it is being resolved by `Promise`. Is it right?

Reply:
I don't think so, specifically because of the comment I quoted, `"If the promise is resolved, its result will be dispatched as an action."`.

The action is passed through the promise, but the "`redux action`" itself doesn't happen until store.dispatch is called. So the action is still sync. I guess this is sort of a semantics question more than anything, though.

Reply:
Oh, I see. 
This discussion is helping me to understand redux a little better.
Please, correct if I'm wrong, but what is dispatched is the result of an action function, for example `{ type: REMOVE_LAND_FOR_SALE, landForSale }` so, we can write an async action function that returns this object and do the same check as the example (if the action function is or an async or not). Make sense?

Reply:
I'm using this definition of action. "Actions are payloads of information that send data from your application to your store."

> Please, correct if I'm wrong, but what is dispatched is the result of an action function

So I would say it like this: The dispatch of the action can be delayed until receiving the result of an async function with middleware written one way. Alternatively, an action can be dispatched by a wrapper function that also triggers an async function after the action has been dispatched (and I think that's what we're doing as it's written here).

> so, we can write an async action function that returns this object and do the same check as the example (if the action function is or an async or not). Make sense?

I don't think we can write an async function that handles actions, but I wouldn't call it an "action function".

"Action creators are exactly that—functions that create actions. It's easy to conflate the terms “action” and “action creator”, so do your best to use the proper term." Source: https://redux.js.org/basics/actions

We definitely can check if the action creator function is async or not.

Reply:
Oh.. great thanks! I was misunderstanding `action` x `action creator`. It's all clear now.

Reply:
Great!

Reply:
The way redux uses action is actually quite similar to the way we define it. A thing that can be sent to the data store right away or later (redux vs. Ethereum blockchain) which is a state mutation.